File: tests/test_partition_logic.rs
use std::path::{Path, PathBuf};
use s2mflow::{load_min_instance, generate_multi_commodity_data};

#[test]
fn test_partition_logic() {
    let data_dir_env = std::env::var("S2MFLOW_DATA_DIR").unwrap_or_else(|_| "data".to_string());
    let data_dir = PathBuf::from(data_dir_env);

    if !data_dir.exists() {
        println!("cargo:warning=Data directory not found.");
        return;
    }

    let mut min_files = Vec::new();
    fn find_min_files_recursive(dir: &Path, files: &mut Vec<PathBuf>) {
        if let Ok(entries) = std::fs::read_dir(dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_dir() {
                    find_min_files_recursive(&path, files);
                } else if path.extension().map_or(false, |ext| ext == "min") {
                    files.push(path);
                }
            }
        }
    }

    find_min_files_recursive(&data_dir, &mut min_files);

    assert!(!min_files.is_empty(), "No .min instances found inside s2mflow_data.");

    let num_commodities_space = vec![2, 3, 5, 10];
    let is_uniform_space = vec![true, false];
    let seed = 42;

    let mut failed_cases = Vec::new();

    for file_path in &min_files {
        let file_name = file_path.file_name().unwrap().to_str().unwrap().to_string();
        let network = load_min_instance(file_path.to_str().unwrap().to_string()).unwrap();

        for &num_commodities in & num_commodities_space {
            for &is_uniform in &is_uniform_space {

                let case_context = format!(
                    "File: {}, K: {}, Uniform: {}",
                    file_name, num_commodities, is_uniform
                );

                println!("Testing Partition Logic: {}", case_context);

                let result = std::panic::catch_unwind(|| {
                    let mc_data = generate_multi_commodity_data(
                        &network, 
                        num_commodities, 
                        is_uniform, 
                        false, 
                        1.0, 
                        1.0, 
                        false, 
                        1.0, 
                        1.0, 
                        seed
                    );

                    // Local Balance and Sign Consistency Checks
                    for (&node_id, partitioned_vals) in &mc_data.supply_partition {
                        let original_supply = *network.supplies.get(&node_id).unwrap_or(&0i64);
                        
                        // Local Partitioning: sum_k(b_i^k) == b_i for each i
                        let local_sum: i64 = partitioned_vals.iter().sum();
                        assert_eq!(
                            local_sum, 
                            original_supply,
                            "Local balance broken at node {}. Got sum={}, expected={} ({})",
                            node_id, local_sum, original_supply, case_context
                        );
                        
                        for &val in partitioned_vals {
                            if original_supply > 0 {
                                assert!(
                                    val >= 0,
                                    "Sign violation: Postive supply node {} generated negative partition value {} ({}).",
                                    node_id, val, case_context
                                );
                            } else if original_supply < 0 {
                                assert!(
                                    val <= 0,
                                    "Sign violation: Negative demand node {} generated positive partition value {} ({}).",
                                    node_id, val, case_context
                                );
                            }
                        }
                    }

                    // Global Balance Check: sum_i(b_i^k) == 0 for rach k.
                    for k in 0..num_commodities {
                        let mut commodity_global_sum: i64 = 0;

                        for node_vals in mc_data.supply_partition.values() {
                            commodity_global_sum += node_vals[k];
                        }

                        assert_eq!(
                            commodity_global_sum, 0,
                            "Global conservation borken for commodity index {}! Summed nodes total was {} ({}).",
                            k, commodity_global_sum, case_context
                        );
                    }

                });

                if result.is_err() {
                    failed_cases.push(case_context);
                }

            }
        }

    }

    if !failed_cases.is_empty() {
        println!("\n {} Partition Logic Test Failed:", failed_cases.len());
        for case in &failed_cases {
            println!("  > {}", case);
        }
        panic!("Partition logic testing suite failed.");
    } else {
        println!("\n Partition logic test passed perfectly.")
    }

}

File: tests/test_partition_logic.py
import os
import glob
import pytest
import s2mflow


DATA_DIR = os.getenv("S2MFLOW_DATA_DIR", "data")

def discover_real_min_files():
    search_pattern = os.path.join(DATA_DIR, "**", "*.min")
    found_files = glob.glob(search_pattern, recursive=True)

    allowed_instances = {"netgen_1.min"}
    filtered_files = [
        f for f in found_files
        if os.path.basename(f) in allowed_instances
    ]

    return sorted([os.path.abspath(f) for f in filtered_files])


REAL_MIN_INSTANCES = discover_real_min_files()

@pytest.mark.skipif(not REAL_MIN_INSTANCES, reason="No .min instances found inside data.")
@pytest.mark.parametrize("input_min_file", REAL_MIN_INSTANCES)
@pytest.mark.parametrize("is_uniform", [True, False])
@pytest.mark.parametrize("num_commodities", [2, 3, 5, 10])
def test_partition_logic(input_min_file, is_uniform, num_commodities):
    """
    Verifies that the multicommody instance generation is correct:
    1. Sign Consistency
    2. Local Partitioning: sum_k(b_i^k) == b_i for each i.
    3. Global Balance: sum_i(b_i^k) == 0 for each k.
    """
    base_name = os.path.basename(input_min_file)

    # 1. Load the single-commodity instance
    network = s2mflow.load_min_instance(input_min_file)

    seed = 42

    mc_data = s2mflow.generate_multi_commodity_data(
        network,
        num_commodities=num_commodities,
        is_uniform=is_uniform,
        randomize_caps=False,
        cap_a=1.0,
        cap_b=1.0,
        randomize_costs=False,
        cost_a=1.0,
        cost_b=1.0,
        seed=seed,
    )

    for node_id, partitioned_vals in mc_data.supply_partition.items():
        original_supply = network.supplies.get(node_id, 0)

        # Local Balance
        local_sum = sum(partitioned_vals)
        assert local_sum == original_supply, \
            f"Local balance broken at node {node_id} in {base_name} (Uniform={is_uniform}, K={num_commodities})." \
            f"Got sum={local_sum}, expected={original_supply}"
        
        # Sign Consistency
        for val in partitioned_vals:
            if original_supply > 0:
                assert val >= 0, \
                    f"Sign violation in {base_name}: Supply node {node_id} with: {val} (Uniform={is_uniform}, K={num_commodities}, org={original_supply})."
            elif original_supply < 0:
                assert val <= 0, \
                    f"Sign violation in {base_name}: Demand node {node_id} with: {val} (Uniform={is_uniform}, K={num_commodities}, org={original_supply})."

    # Global Balance
    for k in range(num_commodities):
        commodity_global_sum = sum(node_vals[k] for node_vals in mc_data.supply_partition.values())
        assert commodity_global_sum == 0, \
            f"Global balance broken for commodity {k} in {base_name} (Uniform={is_uniform}, K={num_commodities})."


    